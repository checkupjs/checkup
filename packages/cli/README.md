# @checkup/cli

A CLI that provides health check information about your project.

![CI Build](https://github.com/checkupjs/checkup/workflows/CI%20Build/badge.svg)
[![oclif](https://img.shields.io/badge/cli-oclif-brightgreen.svg)](https://oclif.io)
[![Version](https://img.shields.io/npm/v/@checkup/cli.svg)](https://npmjs.org/package/@checkup/cli)
[![Downloads/week](https://img.shields.io/npm/dw/@checkup/cli.svg)](https://npmjs.org/package/@checkup/cli)
[![License](https://img.shields.io/npm/l/@checkup/cli.svg)](https://github.com/checkupjs/checkup/blob/master/package.json)

<!-- toc -->

- [Usage](#usage)
- [Configuration](#configuration)
- [Commands](#commands)
  <!-- tocstop -->

# Usage

<!-- usage -->

```sh-session
$ yarn global add @checkup/cli

$ checkup COMMAND
running command...

$ checkup (-v|--version|version)
@checkup/cli/0.0.0 darwin-x64 node-v10.18.0

$ checkup --help [COMMAND]
USAGE
  $ checkup COMMAND
...
```

<!-- usagestop -->

# Configuration

<!-- configuration -->

checkup is designed to be completely configurable via a configuration object.

checkup uses [cosmiconfig](https://github.com/davidtheclark/cosmiconfig) to find and load your configuration object. Starting from the current working directory, it looks for the following possible sources:
                                        
- a checkup property in package.json
- a .checkuprc file
- a checkup.config.js file exporting a JS object

The search stops when one of these is found, and checkup uses that object. 

The .checkuprc file (without extension) can be in JSON or YAML format. You can add a filename extension to help your text editor provide syntax checking and highlighting:

- checkup.json
- checkup.yaml / .checkup.yml
- checkup.js

The configuration object has the following properties:

## plugins

checkup supports the use of plugins. Before using the plugin, you have to install it using npm/yarn.

```sh-session
$ yarn add -D @checkup/plugin-ember
```

To configure plugins, use the plugins key in your configuration file, which contains a list of plugin names.
```json
{
  "plugins": [
    "@checkup/plugin-ember"
  ]
}
```
<!-- TODO: Describe properties in CheckupConfig -->

<!-- configurationstop -->

# Commands

<!-- commands -->

<!-- commandsstop -->
