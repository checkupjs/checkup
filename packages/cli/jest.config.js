module.exports = {
  testEnvironment: 'node',
  moduleFileExtensions: ['ts', 'js', 'json'],
  transform: { '\\.ts$': 'ts-jest' },
  coverageReporters: ['lcov', 'text-summary'],
  // collectCoverage: !!process.env.CI,
  collectCoverageFrom: ['src/**/*.ts'],
  coveragePathIgnorePatterns: ['/templates/'],
  coverageThreshold: {
    global: {
      branches: 100,
      functions: 100,
      lines: 100,
      statements: 100,
    },
  },
  testPathIgnorePatterns: ['/__utils__/'],
};
