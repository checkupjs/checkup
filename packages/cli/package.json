{
  "name": "@checkup/cli",
  "description": "A CLI that provides health check information about your project.",
  "version": "0.0.10",
  "author": "Steve Calvert <steve.calvert@gmail.com>",
  "bin": {
    "checkup": "./bin/run"
  },
  "bugs": "https://github.com/checkupjs/checkup/issues",
  "dependencies": {
    "@babel/traverse": "^7.9.5",
    "@babel/types": "7.9.5",
    "@checkup/core": "0.0.10",
    "@oclif/command": "^1",
    "@oclif/config": "^1",
    "@oclif/plugin-help": "^2",
    "chalk": "^4.0.0",
    "checkup-plugin-ember": "0.0.10",
    "checkup-plugin-ember-octane": "0.0.10",
    "chrome-debugging-client": "^1.2.0",
    "date-and-time": "^0.13.1",
    "debug": "^4.1.1",
    "devtools-protocol": "^0.0.761827",
    "fs-extra": "^9.0.0",
    "handlebars-helpers": "^0.10.0",
    "json-stable-stringify": "^1.0.1",
    "lodash": "^4.17.15",
    "p-map": "^4.0.0",
    "promise.hash.helper": "^1.0.7",
    "recast": "^0.19.0",
    "shorthash": "^0.0.2",
    "tmp": "^0.2.0",
    "tslib": "^1",
    "yeoman-environment": "^2.9.6",
    "yeoman-generator": "^4.9.0"
  },
  "devDependencies": {
    "@checkup/test-helpers": "0.0.10",
    "@oclif/dev-cli": "^1",
    "@oclif/test": "^1",
    "@types/yeoman-environment": "^2.3.3",
    "@types/yeoman-generator": "^3.1.4",
    "yeoman-test": "^2.4.1"
  },
  "engines": {
    "node": ">= 12.*"
  },
  "files": [
    "/bin",
    "/lib",
    "/templates",
    "/static"
  ],
  "homepage": "https://github.com/checkupjs/checkup",
  "keywords": [
    "oclif"
  ],
  "license": "MIT",
  "main": "lib/index.js",
  "oclif": {
    "commands": "./lib/commands",
    "bin": "checkup"
  },
  "publishConfig": {
    "access": "public"
  },
  "repository": "https://github.com/checkupjs/checkup",
  "scripts": {
    "build": "tsc --build",
    "prepack": "yarn build && oclif-dev readme",
    "test": "jest --runInBand --no-cache",
    "version": "oclif-dev readme && git add README.md"
  },
  "types": "lib/index.d.ts"
}
