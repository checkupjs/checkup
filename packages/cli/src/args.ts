export const pathArg = {
  name: 'paths',
  description:
    'The paths or globs that checkup will operate on. If no paths or globs are provided, checkup will run on the entire directory beginning at --cwd.',
};
