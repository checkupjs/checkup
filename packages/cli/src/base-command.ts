import { CheckupError } from '@checkup/core';
import { Command } from '@oclif/command';
import { red } from 'chalk';

export abstract class BaseCommand extends Command {
  extendedError(
    error: string | Error | CheckupError,
    options: { code?: string; exit?: number | false } = {}
  ) {
    if (typeof error === 'string') {
      error = new Error(error);
    }

    if (options.exit === false) {
      console.error(
        error instanceof CheckupError ? error.render() : `\n${red('Error')} ${error.message}`
      );
    } else {
      throw error;
    }
  }
}
