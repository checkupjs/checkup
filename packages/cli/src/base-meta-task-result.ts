import { TaskIdentifier } from '@checkup/core';

export default abstract class BaseMetaTaskResult {
  meta: TaskIdentifier;

  constructor(meta: TaskIdentifier) {
    this.meta = meta;
  }
}
