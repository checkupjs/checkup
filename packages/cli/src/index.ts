import * as path from 'path';

import { Command, flags } from '@oclif/command';
import {
  TaskConstructor,
  TaskList,
  TaskResult,
  getPackageJson,
  getRegisteredTasks,
  ui,
} from '@checkup/core';

function mergeTaskResults(taskResults: TaskResult[]) {
  let mergedResults = {};

  taskResults.forEach(taskResult => {
    mergedResults = Object.assign(mergedResults, taskResult.toJson());
  });

  return mergedResults;
}

class Checkup extends Command {
  static description = 'A CLI that provides health check information about your project';

  static args = [
    {
      name: 'path',
      required: true,
      description: 'The path referring to the root directory that Checkup will run in',
      default: '.',
    },
  ];

  static flags = {
    version: flags.version({ char: 'v' }),
    help: flags.help({ char: 'h' }),
    force: flags.boolean({ char: 'f' }),
    silent: flags.boolean({ char: 's' }),
    json: flags.boolean(),
    task: flags.string({ char: 't' }),
  };

  async run() {
    let { args, flags } = this.parse(Checkup);
    let registeredTasks: Map<string, TaskConstructor>;

    try {
      getPackageJson(args.path);
    } catch (e) {
      this.error(
        `The ${path.resolve(
          args.path
        )} directory found through the 'path' option does not contain a package.json file. You must run checkup in a directory with a package.json file.`,
        e
      );
    }

    await this.config.runHook('register-tasks', {});

    registeredTasks = getRegisteredTasks();

    let taskResults;
    let tasksToBeRun = new TaskList();

    if (flags.task !== undefined) {
      let task = registeredTasks.get(flags.task);

      if (task !== undefined) {
        tasksToBeRun.addTask(task, args);
      }
    } else {
      tasksToBeRun.addTasks(Array.from(registeredTasks.values()), args);
    }

    ui.action.start('Checking up on your project');

    taskResults = await tasksToBeRun.runTasks();

    if (!flags.silent) {
      if (flags.json) {
        ui.styledJSON(mergeTaskResults(taskResults));
      } else {
        taskResults.forEach(taskResult => taskResult.toConsole());
      }
    }

    ui.action.stop();
  }
}

export = Checkup;
