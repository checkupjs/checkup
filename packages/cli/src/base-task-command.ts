import { Result, Log, Invocation } from 'sarif';
import {
  RunFlags,
  TaskError,
  Action,
  CheckupConfig,
  Task,
  TaskContext,
  ui,
  CheckupError,
  getRegisteredActions,
  getConfigPathFromOptions,
  getConfigPath,
  readConfig,
  registerParser,
  registerActions,
  registerTaskReporter,
  loadPlugins,
  getRegisteredParsers,
  FilePathArray,
  getFilePathsAsync,
  buildNotificationsFromTaskErrors,
} from '@checkup/core';
import { BaseCommand } from './base-command';
import TaskList from './task-list';
import { getLog } from './get-log';
import { TASK_ERRORS } from './task-errors';
import { getPackageJson } from './utils/get-package-json';
import { getReporter } from './reporters/get-reporter';
import { reportAvailableTasks } from './reporters/available-task-reporter';

let __tasksForTesting: Set<Task> = new Set<Task>();

export function _registerTaskForTesting(task: Task) {
  __tasksForTesting.add(task);
}
export function _resetTasksForTesting() {
  __tasksForTesting = new Set<Task>();
}

export abstract class BaseTaskCommand extends BaseCommand {
  runArgs!: string[];
  runFlags!: RunFlags;
  pluginTasks: TaskList = new TaskList();
  pluginTaskResults: Result[] = [];
  pluginTaskErrors: TaskError[] = [];
  startTime: string = '';
  actions: Action[] = [];
  checkupConfig!: CheckupConfig;
  executedTasks!: Task[];
  cliModeEnabled: boolean = true;
  taskContext!: TaskContext;

  get taskFilterType() {
    if (this.runFlags.task !== undefined) {
      return 'task';
    } else if (this.runFlags.category !== undefined) {
      return 'category';
    } else if (this.runFlags.group !== undefined) {
      return 'group';
    }

    return '';
  }

  public async run() {
    await this.loadConfig();

    await this.register();

    if (this.runFlags['list-tasks']) {
      this.printAvailableTasks();
    } else {
      if (this.cliModeEnabled) {
        ui.action.start('Checking up on your project');
      }

      await this.runTasks();
      await this.runActions();

      let log: Log = await getLog(
        this.taskContext,
        this.pluginTaskResults,
        this.actions,
        this.getInvocation(this.pluginTaskErrors),
        this.pluginTasks,
        this.executedTasks
      );

      if (this.cliModeEnabled) {
        this.report(log);
        ui.action.stop();
      }

      return log;
    }
  }

  private async runTasks() {
    if (this.taskFilterType) {
      let { tasksFound, tasksNotFound } = this.findTasks();

      if (tasksFound.length > 0) {
        [this.pluginTaskResults, this.pluginTaskErrors] = await this.pluginTasks.runTasks(
          tasksFound
        );
      }

      if (tasksNotFound.length > 0) {
        let error = TASK_ERRORS.get(this.taskFilterType)!;

        this.extendedError(new CheckupError(error.message(tasksNotFound), error.callToAction));
        ui.action.stop();
      }

      this.executedTasks = tasksFound;
    } else {
      [this.pluginTaskResults, this.pluginTaskErrors] = await this.pluginTasks.runTasks();
      this.executedTasks = this.pluginTasks.getTasks();
    }
  }

  private findTasks() {
    if (this.runFlags.task !== undefined) {
      return this.pluginTasks.findAllByTaskName(...this.runFlags.task);
    } else if (this.runFlags.category !== undefined) {
      return this.pluginTasks.findAllByCategory(...this.runFlags.category);
    } else if (this.runFlags.group !== undefined) {
      return this.pluginTasks.findAllByGroup(...this.runFlags.group);
    }

    return { tasksFound: [], tasksNotFound: [] };
  }

  private runActions() {
    return new Promise<void>((resolve, reject) => {
      let evaluators = getRegisteredActions();

      for (let [taskName, evaluator] of evaluators) {
        let task = this.pluginTasks.find(taskName);

        let taskResult = this.pluginTaskResults.filter((result: Result) => {
          return result.properties?.taskName === taskName;
        });

        if (task && taskResult.length > 0) {
          try {
            this.actions.push(...evaluator(taskResult, task.config));
          } catch (error) {
            reject(error);
          }
        }
      }

      resolve();
    });
  }

  private async loadConfig() {
    let configPath;

    try {
      configPath =
        (await getConfigPathFromOptions(this.runFlags.config)) || getConfigPath(this.runFlags.cwd);
      this.checkupConfig = readConfig(configPath);

      let plugins = await loadPlugins(this.checkupConfig.plugins, this.runFlags.cwd);

      this.config.plugins.push(...plugins);
    } catch (error) {
      this.extendedError(error);
    }
  }

  protected async register() {
    await this.config.runHook('register-parsers', {
      registerParser,
    });

    await this.config.runHook('register-actions', {
      registerActions,
    });

    await this.config.runHook('register-task-reporter', {
      registerTaskReporter,
    });

    // if excludePaths are provided both via the command line and config, the command line is prioritized
    let excludePaths = this.runFlags['exclude-paths'] || this.checkupConfig.excludePaths;

    this.taskContext = Object.freeze({
      cliArguments: this.runArgs,
      cliFlags: this.runFlags,
      parsers: getRegisteredParsers(),
      config: this.checkupConfig,
      pkg: getPackageJson(this.runFlags.cwd),
      paths: FilePathArray.from(
        await getFilePathsAsync(this.runFlags.cwd, this.runArgs, excludePaths)
      ) as FilePathArray,
    });

    this.registerTasks();

    __tasksForTesting.forEach((task: Task) => {
      this.pluginTasks.registerTask(task);
    });
  }

  protected abstract registerTasks(): void;

  private report(log: Log) {
    let generateReport = getReporter(this.runFlags);

    generateReport(log, this.runFlags);
  }

  private printAvailableTasks() {
    reportAvailableTasks(this.pluginTasks);
  }

  private getInvocation(errors: TaskError[]): Invocation {
    return {
      arguments: this.runArgs,
      executionSuccessful: true,
      endTimeUtc: new Date().toJSON(),
      environmentVariables: {
        cwd: this.runFlags.cwd,
        outputFile: this.runFlags['output-file'],
        format: this.runFlags.format,
      },
      toolExecutionNotifications: buildNotificationsFromTaskErrors(errors),
      startTimeUtc: this.startTime,
    };
  }
}
