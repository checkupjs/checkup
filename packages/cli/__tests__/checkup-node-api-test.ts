import { Checkup } from '@checkup/cli';
import { CheckupProject } from '@checkup/test-helpers';
import { _registerTaskForTesting, _resetTasksForTesting } from '../src/base-task-command';

const TEST_TIMEOUT = 100000;

describe('@checkup/cli', () => {
  describe('node API output', () => {
    let project: CheckupProject;

    beforeEach(function () {
      project = new CheckupProject('checkup-app', '0.0.0', (project) => {
        project.addDependency('react', '^15.0.0');
        project.addDependency('react-dom', '^15.0.0');
      });
      project.files['index.js'] = 'module.exports = {};';
      project.files['index.hbs'] = '<div>Checkup App</div>';

      project.addCheckupConfig({
        plugins: [],
        tasks: {},
      });
      project.writeSync();
      project.gitInit();
      project.install();
    });

    afterEach(function () {
      project.dispose();
    });

    it(
      'should output checkup result',
      async () => {
        let result = await Checkup.run(['--cwd', project.baseDir]);

        expect(result).toMatchSnapshot({
          runs: expect.any(Array),
        });
      },
      TEST_TIMEOUT
    );
  });
});
