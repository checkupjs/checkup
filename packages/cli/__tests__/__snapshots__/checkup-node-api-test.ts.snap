// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@checkup/cli node API output should output checkup result 1`] = `
Object {
  "$schema": "https://schemastore.azurewebsites.net/schemas/json/sarif-2.1.0-rtm.5.json",
  "properties": Object {
    "actions": Array [],
    "analyzedFiles": FilePathArray [
      ".checkuprc",
      ".eslintrc.js",
      "index.hbs",
      "index.js",
      "package.json",
      "yarn.lock",
    ],
    "analyzedFilesCount": 6,
    "cli": Object {
      "args": Object {
        "paths": Array [],
      },
      "config": Object {
        "$schema": "https://raw.githubusercontent.com/checkupjs/checkup/master/packages/core/src/schemas/config-schema.json",
        "excludePaths": Array [],
        "plugins": Array [],
        "tasks": Object {},
      },
      "configHash": "dd17cda1fc2eb2bc6bb5206b41fc1a84",
      "flags": Object {
        "config": undefined,
        "excludePaths": undefined,
        "format": "stdout",
        "outputFile": "",
        "task": undefined,
      },
      "schema": 1,
      "version": "0.0.0",
    },
    "project": Object {
      "name": "checkup-app",
      "repository": Object {
        "activeDays": "0 days",
        "age": "0 days",
        "linesOfCode": Object {
          "total": 3,
          "types": Array [
            Object {
              "extension": "js",
              "total": 2,
            },
            Object {
              "extension": "hbs",
              "total": 1,
            },
          ],
        },
        "totalCommits": 0,
        "totalFiles": 0,
      },
      "version": "0.0.0",
    },
    "timings": Object {},
  },
  "runs": Any<Array>,
  "version": "2.1.0",
}
`;
